selfdrive/car/hyundai/carstate.py

```diff
@@ -0,0 +1,183 @@
+
+    self.park_brake = False
+    self.main_on = True
+    self.acc_active = cp.vl["SCC12"]['ACCMode'] == 1
+    self.pcm_acc_status = int(self.acc_active)
+
+    # calc best v_ego estimate, by averaging two opposite corners
+    self.v_wheel_fl = cp.vl["WHL_SPD11"]['WHL_SPD_FL'] * CV.KPH_TO_MS
+    self.v_wheel_fr = cp.vl["WHL_SPD11"]['WHL_SPD_FR'] * CV.KPH_TO_MS
+    self.v_wheel_rl = cp.vl["WHL_SPD11"]['WHL_SPD_RL'] * CV.KPH_TO_MS
+    self.v_wheel_rr = cp.vl["WHL_SPD11"]['WHL_SPD_RR'] * CV.KPH_TO_MS
+    self.v_wheel = (self.v_wheel_fl + self.v_wheel_fr + self.v_wheel_rl + self.v_wheel_rr) / 4.
+
+    # There is a 2% error on Sorento GT due to slightly larger wheel diameter compared to poverty packs.
+    #   Dash assumes about 4% which is excessive. Using 3% to be safe
+    if self.car_fingerprint == CAR.SORENTO:
+      self.v_wheel = self.v_wheel * 1.03 
+    self.low_speed_lockout = self.v_wheel < 1.0
+
+    # Kalman filter
+    if abs(self.v_wheel - self.v_ego) > 2.0:  # Prevent large accelerations when car starts at non zero speed
+      self.v_ego_x = np.matrix([[self.v_wheel], [0.0]])
+
+    self.v_ego_raw = self.v_wheel
+    v_ego_x = self.v_ego_kf.update(self.v_wheel)
+    self.v_ego = float(v_ego_x[0])
+    self.a_ego = float(v_ego_x[1])
+    self.cruise_set_speed = cp.vl["SCC11"]['VSetDis']
+    self.standstill = not self.v_wheel > 0.001
+
+    self.angle_steers = cp.vl["SAS11"]['SAS_Angle']
+    self.angle_steers_rate = cp.vl["SAS11"]['SAS_Speed']
+    self.yaw_rate = cp.vl["ESP12"]['YAW_RATE']
+    self.main_on = True
+    self.left_blinker_on = cp.vl["CGW1"]['CF_Gway_TSigLHSw']
+    self.left_blinker_flash = cp.vl["CGW1"]['CF_Gway_TurnSigLh']
+    self.right_blinker_on = cp.vl["CGW1"]['CF_Gway_TSigRHSw']
+    self.right_blinker_flash = cp.vl["CGW1"]['CF_Gway_TurnSigRh']
+    self.steer_override = abs(cp.vl["MDPS11"]['CR_Mdps_DrvTq']) > 1.0
+    self.steer_state = cp.vl["MDPS12"]['CF_Mdps_ToiActive'] #0 NOT ACTIVE, 1 ACTIVE
+    self.steer_error = cp.vl["MDPS12"]['CF_Mdps_ToiUnavail']
+    self.brake_error = 0
+    self.steer_torque_driver = cp.vl["MDPS11"]['CR_Mdps_DrvTq'] 
+    self.steer_torque_motor = cp.vl["MDPS12"]['CR_Mdps_OutTq']
+
+    self.user_brake = 0
+
+    self.brake_pressed = cp.vl["TCS13"]['DriverBraking']
+    self.brake_lights = bool(self.brake_pressed)
+    if (cp.vl["TCS13"]["DriverOverride"] == 0 and cp.vl["TCS13"]['ACC_REQ'] == 1):
+      self.pedal_gas = 0
+    else: 
+      self.pedal_gas = cp.vl["EMS12"]['TPS']
+    self.car_gas = cp.vl["EMS12"]['TPS']
+
+    # Gear Selecton - This should be compatible with all Kia/Hyundai with Auto's
+    if cp.vl["CLU15"]["CF_Clu_InhibitD"] == 1:
+      self.gear_shifter = "drive"
+    elif cp.vl["CLU15"]["CF_Clu_InhibitN"] == 1:
+      self.gear_shifter = "neutral"
+    elif cp.vl["CLU15"]["CF_Clu_InhibitP"] == 1:
+      self.gear_shifter = "park"
+    elif cp.vl["CLU15"]["CF_Clu_InhibitR"] == 1:
+      self.gear_shifter = "reverse"
```

Comment:
are the cruise buttons (set/resume) on can?

Reply:
They are, yes.